File: leetcode/python/77_combinations.py
class Solution:
    def combine(self, n: int, k: int) -> list[list[int]]:
        res: list[list[int]] = []
        self.helper(1, n, k, [], res)
        return res
    
    def helper(self, low: int, high: int, k: int, pair: list[int], res: list[list[int]]) -> None:
        if k == 0:
            res.append(pair.copy())
            return
        
        for i in range(low, high + 2 - k):
            pair.append(i)
            self.helper(i+1, high, k-1, pair, res)
            pair.pop()

if __name__ == '__main__':
    s = Solution()
    print(s.combine(4, 2))
    print(s.combine(1, 1))
    print(s.combine(6, 4))

File: leetcode/rust/src/77_combinations.rs
pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    let mut path = Vec::new();
    dfs(n, k, 1, &mut path, &mut res);
    res
}

fn dfs(n: i32, k: i32, start: i32, path: &mut Vec<i32>, res: &mut Vec<Vec<i32>>) {
    if path.len() == k as usize {
        res.push(path.clone());
        return;
    }
    for i in start..=n {
        path.push(i);
        dfs(n, k, i + 1, path, res);
        path.pop();
    }
}